checkers/src/Player.cpp

```diff
@@ -56,47 +63,58 @@ std::vector<std::vector<Move>> Player::getValidMoves(Board &board){
     return valid_moves;
 }
 
-std::vector<Move> Player::getValidMovePiece(Board &board, int index){
-    std::vector<Move> move;
-    move = (pieces_[index])->getCaptureMoves(board);
-    if(move.empty()) (pieces_[index]) ->getValidMoves(board);
-
-    return move;
-}
-
-void Player::printPlayer(){
-    for (int i = 0; i < 12; i++){
+/*void Player::printPlayer(){
+    for (size_t i = 0; i < 12; i++){
         std::cout << pieces_[i]->getPosition().x << "," << pieces_[i]->getPosition().y << std::endl; 
     }
-}
+}*/
 
 void Player::erasePiece(std::shared_ptr<Piece> piece){
     pieces_.erase(std::remove_if(pieces_.begin(), pieces_.end(), 
-                                            [&piece](std::shared_ptr<Piece> p) {return p && p == piece ; }));
+                                            [&piece](const std::shared_ptr<Piece> &p) {return p && p == piece ; }));
 }
 
 void Player::changePiece(std::shared_ptr<Piece> piece, Position pos){
     piece->changePosition(pos);
 }
 
+void Player::addPiece(bool isKing, Position pos, Board &board){
+    if(isKing) pieces_.push_back(std::make_shared<King>(pos, isWhite_, isUser_, board));
+    else pieces_.push_back(std::make_shared<Pawn>(pos, isWhite_, isUser_, board));
+}
+
 void Player::movePiece(Board &board, Player &opponent, Move move){
-    auto tmp = findPiece(move.getStartPosition());
+    auto start = move.getStartPosition();
+    //auto tmp = findPiece(start);
+    auto it = std::find_if(pieces_.begin(), pieces_.end(), 
+                    [&start](const std::shared_ptr<Piece> &p) {return p->getPosition() == start; });
+    if(it == pieces_.end()) throw std::out_of_range("Nie znaleziona pionka na podanej pozycji");
+
     if(!move.getChangedPosition().empty()){
-        erasePiece(tmp);
-        pieces_.push_back(std::make_shared<King>(move.getStartPosition(), isUser_, isWhite_, board));
-        tmp = findPiece(move.getStartPosition());
+        pieces_.erase(it); 
+        //erasePiece(tmp);
+        addPiece(true,start,board);
+        it = std::find_if(pieces_.begin(), pieces_.end(), 
+                    [&start](const std::shared_ptr<Piece> &p) {return p->getPosition() == start; });
+        //tmp = findPiece(start);
+        if(it == pieces_.end()) throw std::out_of_range("Nie znaleziona pionka na podanej pozycji");
```

Comment:
Czemu tu jest jeszcze raz sprawdzane czy jest pionek na startowej pozycji, jeśli linię wcześniej tam stawiamy?

Reply:
to bardziej sprawdzenie, czy addPiece faktycznie dodał tego pionka na pozycji startowej - chociaż to sprawdzenie moze powinno być w addPiece. Szczerze mowiac nie dokonca wiem, ktore rzeczy mamy sprawdzac :D
Chodzilo mi o taka sytuacje, w ktorej addPiece z jakiegos powodu nie stworzyl nowego pionka.

Reply:
Tak myślałem, ale no właśnie, lepiej by to przenieść poziom niżej do addPiece jak już
I know, też mam często problem, co sprawdzać i jak najlepiej rzucać wyjątkami